### PR TITLE
Enforce `check_parameters` for `alpha` in `geometric_adstock`

### PR DESCRIPTION
**Pull request by PabloRoque**
## Description

* Introduces `pymc.distributions.dist_math.check_parameters` to validate `geometric_adstock`'s `alpha`.
  - I took the idea to use that function from pymc's [implementing_distribution.md](implementing_distribution.md)
  - Basically a wrapper around @wd60622 idea of using `pymc.logprob.utils.CheckParameterValue`
* Add a test to assert failure with bad `alpha` values

## Related Issue

- [ ] Closes #
- [x] Related to #918 

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Modules affected

- [x] MMM
- [ ] CLV

## Type of change

- [ ] New feature / enhancement
- [x] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pymc-marketing--960.org.readthedocs.build/en/960/

### CODE REVIEW

File: pymc_marketing/mmm/transformers.py [hunk 1]
```diff
@@ -235,6 +236,11 @@ def geometric_adstock(
        with carryover and shape effects." (2017).
 
     """
+    alpha = check_parameters(
+        alpha, [pt.gt(alpha, 0), pt.lt(alpha, 1)], msg="0 <= alpha <= 1"
+    ).eval()
+    l_max = check_parameters(l_max, [pt.gt(l_max, 0)], msg="l_max >= 0").eval()
```
Comment (wd60622) [inline pymc_marketing/mmm/transformers.py:242]: We won't want to use eval here

Comment (PabloRoque) [reply pymc_marketing/mmm/transformers.py:242]: Yep, I am seeing issues. 

Removing and having a look.

Comment (wd60622) [reply pymc_marketing/mmm/transformers.py:242]: It should only raise when the eval is called on the result of function

```python
# Fine
result = geometric_adstock(np.linspace(0, 1), alpha=-5)
# Causes error
result.eval()

Comment (wd60622) [reply pymc_marketing/mmm/transformers.py:242]: Just like the PyMC distributions: 

```python
# Fine
dist = pm.Normal.dist(mu=0, sigma=-1)
# ValueError
dist.eval()
```

File: tests/mmm/test_transformers.py [hunk 2]
```diff
@@ -131,22 +132,38 @@ def test_geometric_adstock_x_zero(self, mode):
...
+    def test_geometric_adstock_check_parameters(self, x, alpha, l_max, suceeds):
+        if suceeds:
+            y = geometric_adstock(x=x, alpha=alpha, l_max=l_max)
+            y_np = y.eval()
+            assert y_np[0] == x[0]
+            assert y_np[1] == x[1] + alpha * x[0]
+            assert y_np[2] == x[2] + alpha * x[1] + (alpha**2) * x[0]
+        else:
+            with pytest.raises(ParameterValueError):
+                geometric_adstock(x=x, alpha=alpha, l_max=l_max)
```
Comment (wd60622) [inline tests/mmm/test_transformers.py:166]: There should already be some tests for the sucesses. So we can just focus on the behavior of the bad params here

File: pymc_marketing/mmm/transformers.py (hunk 1 above)
Comment (wd60622) [reply pymc_marketing/mmm/transformers.py:242]: Can this check not all be done in the same line? like in the example I linked?

Comment (PabloRoque) [reply pymc_marketing/mmm/transformers.py:242]: I removed the check on l_max > 0:
* If passed, an exception is already raised: `ValueError: length should not be negative`
* `batched_convolution` is called ahead of `.eval()`. So the previous exception is already raised and we never reach `ParameterValueError`

Comment (PabloRoque) [reply pymc_marketing/mmm/transformers.py]: `.eval()` now part of the test as: 

```
y = geometric_adstock(x=x, alpha=alpha, l_max=l_max)
y.eval()
```

File: tests/mmm/test_transformers.py (hunk 2 above)
Comment (PabloRoque) [reply tests/mmm/test_transformers.py]: Left the previous `test_geometric_adstock_good_alpha` as it was. Introduced `test_geometric_adstock_bad_alpha`

File: tests/mmm/test_transformers.py [hunk 3]
```diff
@@ -148,6 +149,21 @@ def test_geometric_adstock_good_alpha(self, x, alpha, l_max):
...
+            (np.ones(shape=(100)), -1.7, 100),
+            (np.ones(shape=(100)), 22.5, 7),
+            (np.linspace(start=0.0, stop=1.0, num=50), -0.8, 3),
+            (np.linspace(start=0.0, stop=1.0, num=50), 8, 50),
+        ],
+    )
+    def test_geometric_adstock_bad_alpha(self, x, alpha, l_max):
+        with pytest.raises(ParameterValueError):
+            y = geometric_adstock(x=x, alpha=alpha, l_max=l_max)
+            y.eval()
```
Comment (wd60622) [inline tests/mmm/test_transformers.py:165]:
```suggestion
        y = geometric_adstock(x=x, alpha=alpha, l_max=l_max)
        with pytest.raises(ParameterValueError):
            y.eval()
```

File: tests/mmm/test_transformers.py [hunk 4]
```diff
@@ -148,6 +149,21 @@ def test_geometric_adstock_good_alpha(self, x, alpha, l_max):
         assert y_np[1] == x[1] + alpha * x[0]
         assert y_np[2] == x[2] + alpha * x[1] + (alpha**2) * x[0]
 
+    @pytest.mark.parametrize(
+        "x, alpha, l_max",
```
Comment (wd60622) [inline tests/mmm/test_transformers.py:153]: Is the `x` important here? can it be the same for all the tests? Would be less moving parts 😄

**Review by wd60622 — COMMENTED** [review]
Think this looks good. Just 2 suggestions

File: pymc_marketing/mmm/transformers.py (hunk 1 above)
Comment (wd60622) [reply pymc_marketing/mmm/transformers.py]: Perf. Don't think we are passing l_max as TensorVariable anyway

File: tests/mmm/test_transformers.py (hunk 4 above)
Comment (PabloRoque) [reply tests/mmm/test_transformers.py:153]: I basically copied the values from `test_geometric_adstock_good_alpha`. 
Let me just test an int and a float smaller and greater than the bounds. `x = np.ones(shape=(100)` for all of them

File: tests/mmm/test_transformers.py [hunk 5]
```diff
@@ -148,6 +149,20 @@ def test_geometric_adstock_good_alpha(self, x, alpha, l_max):
...
+        [
+            (np.ones(shape=(100)), -0.3, 10),
+            (np.ones(shape=(100)), -2, 10),
+            (np.ones(shape=(100)), 22.5, 10),
+            (np.ones(shape=(100)), 2, 10),
+        ],
+    )
+    def test_geometric_adstock_bad_alpha(self, x, alpha, l_max):
+        y = geometric_adstock(x=x, alpha=alpha, l_max=l_max)
```
Comment (juanitorduz) [inline tests/mmm/test_transformers.py:162]: As x is the same for all cases I think it is better to put `x=np.ones(shape=(100)` inside the `test_geometric_adstock_bad_alpha` and just have 

```python
    @pytest.mark.parametrize(
        "alpha, l_max",
    . . . 
```

**Review by juanitorduz — CHANGES_REQUESTED** [review]
Thanks @PabloRoque ! Just a small request and this one is good to go!

File: tests/mmm/test_transformers.py (hunk 5 above)
Comment (wd60622) [inline tests/mmm/test_transformers.py:162]:
```suggestion
    @pytest.mark.parametrize(
        "alpha",
        [-0.3, -2, 22.5, 2],
        ids=["less_than_zero_0", "less_than_zero_1", "greater_than_one_0", "greater_than_one_1"],
    )
    def test_geometric_adstock_bad_alpha(self, alpha):
        l_max = 10
        x = np.ones(shape=100)
        y = geometric_adstock(x=x, alpha=alpha, l_max=l_max)
```

Comment (juanitorduz) [reply tests/mmm/test_transformers.py:162]: haha or this XD

Comment (wd60622) [reply tests/mmm/test_transformers.py:162]: Easier to see what is happening imo

**Comment by PabloRoque** [comment]
Thanks for the suggestions, and having a look at the PR. 
Indeed cleaner this way

**Comment by wd60622** [comment]
pre-commit.ci autofix

**Review by wd60622 — APPROVED** [review]
Looks good! thanks for the edits

**Pull request merged**